swmm-toolkit/src/swmm/toolkit/solver.i

```diff
@@ -30,23 +39,277 @@
 /* GENERATES DOCUMENTATION */
 %feature("autodoc", "2");
 
-
 /* RENAME FUNCTIONS PYTHON STYLE */
 %include "solver_rename.i"
 
-
 /* INSERTS CUSTOM EXCEPTION HANDLING IN WRAPPER */
 %exception
 {
     $function
     if (result > 0) {
-        char errmsg[90];
-        swmm_getError(errmsg, 90);
-        PyErr_SetString(PyExc_Exception, errmsg);
+        char* errorMsg = NULL;
+        int errorCode = 0;
+        errorCode = swmm_getAPIError(result, &errorMsg);
+        if (errorCode == 0) {
+            PyErr_SetString(PyExc_Exception, errorMsg);
+        }
+        swmm_freeMemory(errorMsg);
         SWIG_fail;
     }
 }
 
+%apply int *OUTPUT {
+    int *year,
+    int *month,
+    int *day,
+    int *hour,
+    int *minute,
+    int *second
+};
+
+%apply double *OUTPUT {
+    double *value
+};
+
+%apply signed char *OUTPUT {
+    signed char *value
+};
+
+%apply char *OUTPUT {
+    char *condition
+};
+
+%cstring_output_allocate(char **OUTCHAR, swmm_freeMemory(*$1));
+
+%apply char **OUTCHAR {
+    char **errorMsg,
+    char **id
+};
+
+/* TYPEMAPS FOR MEMORY MANAGEMNET OF DOUBLE ARRAYS */
+%typemap(in, numinputs=0)double** DOUBLEPOINTER (double* doubleArray){
+   $1 = &doubleArray;
+}
+%typemap(argout) (double** DOUBLEPOINTER) {
+    if (*$1) 
+    {
+      int length = sizeof(*$1)/sizeof(double) + 1;
+      PyObject *o = PyList_New(length);
+      double* doubleArray = *$1;
+      for(int i=0; i<length; i++) {
+        PyList_SetItem(o, i, PyFloat_FromDouble(doubleArray[i]));
+      }
+      $result = SWIG_Python_AppendOutput($result, o);
+      swmm_freeMemory(*$1);
+    }
+}
+
+
+%typemap(in, numinputs=0) SM_NodeStats **DOUBLEPOINTER (SM_NodeStats *temp){
```

Comment:
This approach is serviceable. But I have a nagging feeling that it could probably be done with fewer custom typemaps.

Reply:
An approach that is being used in the EPANET code that I find really useful for defining the typemaps is to qualify variable names in the source by adding the prefixes "in_", "out_", or "in_out_" for inputs, outputs and in&out. Could we do that here? It makes it easier to develop the typmaps.